In [1]:
import pandas as pd
df = pd.read_csv('HMAHCC_COMP.csv')

c:\users\chenson\appdata\local\programs\python\python37\lib\site-packages\IPython\core\interactiveshell.py:3044: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
import numpy as np

In [3]:
opioid_data = df[df['PAY_DAY_SUPPLY_CNT'].notnull()]
opioid_data_grouped = opioid_data.groupby(by='id')

In [4]:
from itertools import chain

def get_LTOT(ID):
    try:
        group = opioid_data_grouped.get_group(ID)
        frame = group[['Days', 'PAY_DAY_SUPPLY_CNT']].drop_duplicates()
        frame = frame[frame['Days']>=0]
        frame['drugs until'] = (frame['Days'] + frame['PAY_DAY_SUPPLY_CNT']).astype(int)
        frame['range'] = frame.apply(lambda x : range(x['Days'].astype(int),x['drugs until'].astype(int)),1)

        concat = concatenated = chain(*list(frame['range']))
        concat = set(concat)

        day_frame = pd.DataFrame(columns = ['Days', 'Has Drug?'])
        day_frame['Days'] = range(max(concat)+1)
        day_frame['Has Drug?'] = (day_frame['Days'].isin(concat))
        
        for n in range(180,len(day_frame)+1):
            frame_slice = day_frame.iloc[n-180:n]
            drug_days = np.sum(frame_slice['Has Drug?'])

            if drug_days >= 162:
                return (True, n-180, n)

        return (False, np.nan, np.nan)
    except:
        return (np.nan, np.nan, np.nan)

In [5]:
def get_LTOT_frame(ID):
    try:
        group = opioid_data_grouped.get_group(ID)
        frame = group[['Days', 'PAY_DAY_SUPPLY_CNT']].drop_duplicates()
        frame = frame[frame['Days']>=0]
        frame['drugs until'] = (frame['Days'] + frame['PAY_DAY_SUPPLY_CNT']).astype(int)
        frame['range'] = frame.apply(lambda x : range(x['Days'].astype(int),x['drugs until'].astype(int)),1)

        concat = concatenated = chain(*list(frame['range']))
        concat = set(concat)

        day_frame = pd.DataFrame(columns = ['Days', 'Has Drug?'])
        day_frame['Days'] = range(max(concat)+1)
        day_frame['Has Drug?'] = (day_frame['Days'].isin(concat))
        
        return day_frame
    except:
        return "No data"

## Run the map for response variable:

In [6]:
id_list = df['id'].drop_duplicates().values
id_list

array(['ID10010854159', 'ID10013863216', 'ID10024447278', ...,
       'ID99985278397', 'ID99987501651', 'ID99994197161'], dtype=object)

In [7]:
len(id_list)

14000

In [8]:
%%time
response_variable = pd.DataFrame(id_list, columns = ['id'])
response_variable['LTOT'] = response_variable['id'].map(get_LTOT)
response_variable[['LTOT', 'Begining Date', 'End Date']] = pd.DataFrame(response_variable['LTOT'].tolist(), index = response_variable.index)

Wall time: 9min 38s


In [9]:
response_variable

id   LTOT  Begining Date  End Date
0      ID10010854159  False            NaN       NaN
1      ID10013863216   True            0.0     180.0
2      ID10024447278   True            0.0     180.0
3       ID1002482139   True            0.0     180.0
4       ID1003386406  False            NaN       NaN
5      ID10036289882  False            NaN       NaN
6      ID10036954381  False            NaN       NaN
7      ID10055739763   True          224.0     404.0
8      ID10074598346  False            NaN       NaN
9      ID10081072715  False            NaN       NaN
10      ID1008136105   True          292.0     472.0
11     ID10085004805  False            NaN       NaN
12     ID10085392189   True          247.0     427.0
13     ID10112532931   True            0.0     180.0
14     ID10119459922  False            NaN       NaN
15     ID10125698107   True            0.0     180.0
16     ID10126196784  False            NaN       NaN
17      ID1013630023   True            0.0     180.0
18     ID10137326880   True            0.0     180.0
19     ID10144726843  False            NaN       NaN
20     ID10155139868  False            NaN       NaN
21     ID10158166620  False            NaN       NaN
22     ID10166893764   True            0.0     180.0
23     ID10167700196   True            0.0     180.0
24     ID10169051312   True          389.0     569.0
25     ID10180352120   True            0.0     180.0
26     ID10189080475  False            NaN       NaN
27     ID10194867854  False            NaN       NaN
28     ID10201747254   True          228.0     408.0
29     ID10203805803   True            0.0     180.0
...              ...    ...            ...       ...
13970  ID99819880491   True            0.0     180.0
13971   ID9982180832  False            NaN       NaN
13972  ID99824085831  False            NaN       NaN
13973  ID99827357428   True            0.0     180.0
13974  ID99832270387   True          215.0     395.0
13975  ID99844174319  False            NaN       NaN
13976  ID99851961387  False            NaN       NaN
13977  ID99853738490   True            0.0     180.0
13978  ID99865372013  False            NaN       NaN
13979  ID99866353021  False            NaN       NaN
13980  ID99870677385   True          307.0     487.0
13981  ID99877095827  False            NaN       NaN
13982  ID99888932285  False            NaN       NaN
13983  ID99893580330   True            0.0     180.0
13984  ID99899080861   True          324.0     504.0
13985  ID99899156391  False            NaN       NaN
13986  ID99906098656  False            NaN       NaN
13987  ID99906998779   True            0.0     180.0
13988  ID99915109341   True            0.0     180.0
13989  ID99915340915   True            0.0     180.0
13990  ID99922543950  False            NaN       NaN
13991  ID99932051589  False            NaN       NaN
13992   ID9993577054  False            NaN       NaN
13993  ID99940102826  False            NaN       NaN
13994  ID99952965555  False            NaN       NaN
13995  ID99971049558  False            NaN       NaN
13996  ID99975648894   True            0.0     180.0
13997  ID99985278397   True            0.0     180.0
13998  ID99987501651   True            0.0     180.0
13999  ID99994197161   True            0.0     180.0

[14000 rows x 4 columns]

In [10]:
response_variable['LTOT'].value_counts()

True     7025
False    6923
Name: LTOT, dtype: int64

## Example of what this is doing:

In [11]:
get_LTOT('ID10055739763')

(True, 224, 404)

This means that the first 180 stretch  with opiod supply days is days 224 - 403

In [12]:
get_LTOT_frame('ID10055739763')

Days  Has Drug?
0       0       True
1       1       True
2       2       True
3       3       True
4       4       True
5       5      False
6       6      False
7       7      False
8       8      False
9       9      False
10     10      False
11     11      False
12     12      False
13     13      False
14     14      False
15     15      False
16     16      False
17     17      False
18     18      False
19     19      False
20     20      False
21     21      False
22     22      False
23     23      False
24     24      False
25     25      False
26     26      False
27     27      False
28     28      False
29     29      False
..    ...        ...
647   647       True
648   648       True
649   649       True
650   650       True
651   651       True
652   652       True
653   653       True
654   654       True
655   655       True
656   656       True
657   657       True
658   658       True
659   659       True
660   660       True
661   661       True
662   662       True
663   663       True
664   664       True
665   665       True
666   666       True
667   667       True
668   668       True
669   669       True
670   670       True
671   671       True
672   672       True
673   673       True
674   674       True
675   675       True
676   676       True

[677 rows x 2 columns]

In [13]:
get_LTOT_frame('ID10055739763').iloc[404-180:404]

Days  Has Drug?
224   224      False
225   225      False
226   226      False
227   227      False
228   228      False
229   229      False
230   230      False
231   231      False
232   232      False
233   233      False
234   234      False
235   235      False
236   236      False
237   237      False
238   238      False
239   239      False
240   240      False
241   241       True
242   242       True
243   243       True
244   244       True
245   245       True
246   246       True
247   247       True
248   248       True
249   249       True
250   250       True
251   251       True
252   252       True
253   253       True
..    ...        ...
374   374       True
375   375       True
376   376       True
377   377       True
378   378       True
379   379       True
380   380       True
381   381       True
382   382       True
383   383       True
384   384       True
385   385       True
386   386       True
387   387       True
388   388       True
389   389       True
390   390       True
391   391       True
392   392       True
393   393       True
394   394       True
395   395       True
396   396       True
397   397       True
398   398       True
399   399       True
400   400       True
401   401       True
402   402       True
403   403       True

[180 rows x 2 columns]

In [14]:
np.sum(get_LTOT_frame('ID10055739763').iloc[404-180:404]['Has Drug?'])

162